jlm-hls/src/jlm-hls.cpp

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2022 Magnus Sjalander <work@sjalander.com>
+ * See COPYING for terms of redistribution.
+ */
+
+#include <jive/view.hpp>
+
+#include <jlm/backend/llvm/jlm2llvm/jlm2llvm.hpp>
+#include <jlm/backend/llvm/rvsdg2jlm/rvsdg2jlm.hpp>
+#include <jlm/backend/hls/rvsdg2rhls/rvsdg2rhls.hpp>
+#include <jlm/backend/hls/rhls2firrtl/dot-hls.hpp>
+#include <jlm/backend/hls/rhls2firrtl/firrtl-hls.hpp>
+#include <jlm/backend/hls/rhls2firrtl/mlirgen.hpp>
+#include <jlm/backend/hls/rhls2firrtl/verilator-harness-hls.hpp>
+#include <jlm/frontend/llvm/InterProceduralGraphConversion.hpp>
+#include <jlm/frontend/llvm/LlvmModuleConversion.hpp>
+#include <jlm/ir/RvsdgModule.hpp>
+
+#include <jlm-hls/cmdline.hpp>
+
+#include <llvm/IR/LLVMContext.h>
+#include <llvm/IR/Module.h>
+#include <llvm/IRReader/IRReader.h>
+#include <llvm/Support/raw_os_ostream.h>
+#include <llvm/Support/SourceMgr.h>
+
+static void
+dump_llvm(
+	  jlm::RvsdgModule &module,
+	  std::string outfile)
+{
+	llvm::LLVMContext ctx;
+	jlm::StatisticsDescriptor sd;
+	auto jm = jlm::rvsdg2jlm::rvsdg2jlm(module, sd);
+	auto lm = jlm::jlm2llvm::convert(*jm, ctx);
+	std::error_code EC;
+	llvm::raw_fd_ostream os(outfile, EC);
+	lm->print(os, nullptr);
+}
+
+int
+main(int argc, char ** argv)
+{
+	jlm::cmdline_options flags;
+	parse_cmdline(argc, argv, flags);
+
+	llvm::LLVMContext ctx;
+	llvm::SMDiagnostic err;
+	auto llvmModule = llvm::parseIRFile(flags.ifile.to_str(), err, ctx);
+	llvmModule->setSourceFileName(flags.outfolder.path() + "/jlm_hls");
+	if (!llvmModule) {
+		err.print(argv[0], llvm::errs());
+		exit(1);
+	}
+
+	/* LLVM to JLM pass */
+	auto jlmModule = jlm::ConvertLlvmModule(*llvmModule);
+	auto rvsdgModule = jlm::ConvertInterProceduralGraphModule(
+					*jlmModule,
+					flags.sd);
+
+	if (flags.extract) {
+		auto hlsFunction = jlm::hls::split_hls_function(
+					*rvsdgModule,
+					flags.hls_function);
+
+		dump_llvm(*rvsdgModule, flags.outfolder.path() + "/jlm_hls_rest.ll");
+		dump_llvm(*hlsFunction, flags.outfolder.path() + "/jlm_hls_function.ll");
+	} else if (flags.format == jlm::outputformat::fir) {
+		jlm::hls::rvsdg2rhls(*rvsdgModule);
+
+		std::string output;
+		if (flags.circt) {
+			jlm::hls::MLIRGen hls;
+			output = hls.run(*rvsdgModule);
+		} else {
+			jlm::hls::FirrtlHLS hls;
+			output = hls.run(*rvsdgModule);
+		}
+
+		std::string fileName = flags.outfolder.path() + "/jlm_hls.fir";
+		std::ofstream outFile;
+		outFile.open(fileName);
+		outFile << output;
+		outFile.close();
```

Comment:
Could we put this writing of an string to an output file wrap into its own function?

Reply:
Replaced with stringToFile()